rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+
+## Drawbacks
+
+- It adds more complexity to the language.
+- There are potential performance foot guns when iterating over large
+  collections.
+- The parser and compiler have to become more complex to support this use-case.
+
+## Prior Art
+
+TODO
+
+## Alternatives
+
+### For-Loop
+
+A different approach to iteration is to use a built-in syntax for-loop:
+
+```coffee
+for (key, _value) in . {
+  key = upcase(key)
+}
+```
+
+The biggest strength of this approach is the simplicity of the syntax, and the
+familiarity with many other languages that have for-loops.
+
+It's relevant to mention that this solution also still needs lexical-scoping
+implemented, to avoid "leaking" the values of the `key` and `value` variables
+outside of the loop.
+
+One problem with this approach is that recursive iteration (accessing nested
+object fields) isn't possible, unless we add another special syntax (e.g.
+`recursive for (.., ..) in . {}`). This adds more surface-level syntax and
+removes some of its familiarity, making it a less attractive solution.
+
+An additional problem is that the `key` and `value` variables become "special",
+in that, even though it _appears_ that they aren't used after assignment, the
+`for-loop` expression would actually update the object key after each iteration
+in the loop.
+
+While this is technically the same problem we had to solve in the function-based
+solution, applying that same solution to a `for-loop` again makes it look less
+like for-loops in other languages, defeating one of the strengths of this
+approach:
+
+```coffee
+for (key, value) in . {
+  key = upcase(key)
+
+  (key, value)
+}
+```
+
+A solution to the magic-variable problem would be to allow dynamic paths, and
+have operators directly assign to those paths:
+
+```coffee
+for (key, _value) in . {
+  .[upcase(key)] = value
+}
+```
+
+This solves one problem, but introduces another: using `.<path>` always starts
+at the root of the target. Given the following example:
+
+```json
+{ "foo": { "bar": true } }
+```
+
+How would we use dynamic paths in a recursive for-loop?
+
+```coffee
+recursive for (key, value) in . {
+  .[upcase(key)] = value
+}
+```
+
+Because key is `"foo"` and then `"bar"`, you would end up with:
+
+```json
+{ "FOO": true, "BAR": true }
+```
+
+Which is not the expected outcome.
+
+This could be solved by making `.` relative in the for-loop, but that's a major
+shift from the current way VRL works, requires a new way of accessing the root
+object if you can't use `.`, and goes against the rules as laid out in the
+[design document][doc].
+
+---
+
+- What other approaches have been considered and why did you not choose them?
+- How about not doing this at all?
+
+## Outstanding Questions
+
+- Do we want to introduce any form of lexical-scoping in this RFC, or keep the
+  status-quo for now?
+- Do we want to introduce tuple-expressions or are we satisfied with using
+  a two-element array as the return value for `map`?
```

Comment:
I'm ok with an array as long as the compiler strictly checks this. And I agree, sticking with JSON types is important, otherwise we introduce other issues with mapping tuples to JSON (as you mentioned). I also like @StephenWakely's suggestion of a map as well.

Reply:
- Strictly checking will be there, I'll make a note about it
- Mapping tuples to JSON isn't that strange ([Python can do it](https://pythonexamples.org/run.php?pgm=import+json%0A%0A%23python+tuple%0Amytuple+%3D+%28%22python%22%2C+%22json%22%2C+22%2C+23.04%29%0A%0A%23convert+to+JSON+string%0AjsonStr+%3D+json.dumps%28mytuple%29%0A%0A%23print+json+string%0Aprint%28jsonStr%29%0A%0A%23parse+json%0AjsonArr+%3D+json.loads%28jsonStr%29%0A%0A%23print+third+item+of+the+json+array%0Aprint%28jsonArr%5B2%5D%29)) because JSON allows array elements to be of different types, so in effect, a JSON array is similar to a tuple. Obviously, when deserializing a JSON array to a VRL value, you'd have to choose between a tuple or an array. Having said all that, tuples aren't needed, but they also _can_ be implemented cleanly.
- Using objects is definitely a possibility, it's a bit more verbose (and perhaps _too_ verbose), but it is also more clear, and better fits the rules we list in our design document.